Source/BestProjectEver/Private/Weapons/BPE_Weapon.cpp

```diff
@@ -238,37 +247,48 @@ void ABPE_Weapon::Fire()
 				break;
 			}
 		}
+		
+		ApplyDamage(HitTarget, ShootDirection);
 	}
 }
 
 //----------------------------------------------------------------------------------------------------------------------
-void ABPE_Weapon::TraceUnderCrosshairs(FHitResult& OutHitResult)
+void ABPE_Weapon::ApplyDamage(const FHitResult& HitResult, FVector ShootDirection)
 {
-	const ABPE_PlayerCharacter* PlayerOwner = Cast<ABPE_PlayerCharacter>(OwnerCharacter);
+	AActor* DamagedActor = HitResult.GetActor();
+	if(IsValid(DamagedActor) && IsValid(OwnerCharacter))
+	{
+		UGameplayStatics::ApplyPointDamage(DamagedActor, BaseDamage, ShootDirection, HitResult,
+			OwnerCharacter->GetInstigatorController(),OwnerCharacter, DamageType);	
+	}
+}
 
-	if(IsValid(PlayerOwner))
+//----------------------------------------------------------------------------------------------------------------------
+void ABPE_Weapon::TraceUnderCrosshairs(FHitResult& OutHitResult, FVector& Direction)
```

Comment:
Remember that Out parameters should start with the Out prefix. When I read this in the .h, I assumed you were missing a const since it's not marked as an out parameter

Reply:
Also, it could be a bit more specific. What direction? Maybe something like OutTraceDirection makes it clearer.

Reply:
And just as a comment to let you know, either way is fine. Wherever you have access to the hit result, you could get access to the direction by getting the normalized `TraceEnd - TraceStart`. You could save the extra out parameter that way.

Reply:
![image](https://user-images.githubusercontent.com/85327912/173108627-d936dec2-bf0f-4710-bf0c-29b2547a4a09.png)
Done <3

Reply:
Nice, but you missed the normalization. TraceEnd - TraceStart by itself is not a unit vector

Reply:
![image](https://user-images.githubusercontent.com/85327912/173113082-afafd4fa-9157-4f53-9db3-aa118fc951c8.png)